### PR TITLE
Add matrix logarithm

### PR DESCRIPTION
**Pull request by sylee957**
#### Brief description of what is fixed or changed

This is the first step of implementing matrix logarithm using jordan decomposition method.

But the thing most confusing to me, is that the imaginary values of the principal logarithms for scalars should lie in the interval (−π, π] 
https://en.wikipedia.org/wiki/Complex_logarithm#Definition_of_principal_value
but the imaginary values of the eigenvalues of the principal logarithm for matrices should lie in the interval (−π, π)
https://en.wikipedia.org/wiki/Logarithm_of_a_matrix#Existence

And since sympy sets the `log(-1)` as `I*pi`, should setting the `principal=False` be considered consistent with the sympy's implementation of scalar log?

I don't know if those two wikipedia definitions are consistent each other, regarding a simple case like how `Matrix([[-1]]).log()` should be considered.

#### Release Notes

- matrices
  - Added `Matrix.log` for computing matrix logarithm.

### CODE REVIEW

File: sympy/matrices/matrices.py [hunk 1]
```diff
@@ -3173,6 +3173,122 @@ def _jblock_exponential(b):
...
+        """
+        size = self.rows
+        l = self[0, 0]
+
+        if l.is_zero:
+            raise MatrixError(
+                'Could not take logarithm or reciprocal for the given '
+                'eigenvalue {}'.format(l))
+
+        if l.is_negative and principal:
```
Comment (jksuom) [inline sympy/matrices/matrices.py]: I see no reason why the principal values of matrix logatrithms could not be defined in the same way as for scalars (taking `pi*I` as the imaginary part for negative eigenvalues).

**Comment by oscarbenjamin** [comment]
We need some simplification somewhere in either jordan_form or functions that use the Jordan form or perhaps in matrices in general to stop expression blowup. In this case cancel works nicely:
```julia
In [43]: cancel(exp(Matrix([[x, y], [z, x]]).log()))
Out[43]: 
⎡x  y⎤
⎢    ⎥
⎣z  x⎦
```
The output without cancel is very long though.

Obviously we have to be careful because cancel hangs in this case:
```julia
In [46]: cancel(exp(Matrix([[x, y], [z, t]]).log()))
```

However if we use cancel for each matrix incrementally it can keep up without hanging:
```julia
In [116]: M = Matrix([[x, y], [z, t]])

In [117]: P, J = M.jordan_form()  

In [118]: logM = cancel(cancel(P)*J.log()*cancel(cancel(P).inv()))

In [119]: P2, J2 = logM.jordan_form()

In [120]: cancel(cancel(P2)*exp(J2)*cancel(cancel(P2).inv()))
Out[120]: 
⎡x  y⎤
⎢    ⎥
⎣z  t⎦
```
It's possible that something like cancel used at the element level when multiplying or inverting matrices might be more effective.

It's not just the case with symbols that's slow since numeric expressions can blow up as well. This hangs:
```julia
In [127]: exp(Matrix([[1, 2], [3, 4]]).log(principal=False))
```
(or maybe that's a separate problem)

File: sympy/matrices/matrices.py (hunk 1 above)
Comment (oscarbenjamin) [reply sympy/matrices/matrices.py]: I agree

**Comment by sylee957** [comment]
For numeric cases, I think that the expression blowups can be more likely related if the matrix has square roots for eigenvalues.

```python
>>> from sympy import *
>>> m = Matrix([[1, 2], [3, 4]])

>>> m.exp()
Matrix([
[-2*exp(5/2 - sqrt(33)/2)/((-3/2 + sqrt(33)/2)*(-2/(-3/2 + sqrt(33)/2) + 2/(-sqrt(33)/2 - 3/2))) + 2*exp(5/2 + sqrt(33)/2)/((-sqrt(33)/2 - 3/2)*(-2/(-3/2 + sqrt(33)/2) + 2/(-sqrt(33)/2 - 3/2))), -4*exp(5/2 - sqrt(33)/2)/((-3/2 + sqrt(33)/2)*(-sqrt(33)/2 - 3/2)*(-2/(-3/2 + sqrt(33)/2) + 2/(-sqrt(33)/2 - 3/2))) + 4*exp(5/2 + sqrt(33)/2)/((-3/2 + sqrt(33)/2)*(-sqrt(33)/2 - 3/2)*(-2/(-3/2 + sqrt(33)/2) + 2/(-sqrt(33)/2 - 3/2)))],
[                                                 exp(5/2 - sqrt(33)/2)/(-2/(-3/2 + sqrt(33)/2) + 2/(-sqrt(33)/2 - 3/2)) - exp(5/2 + sqrt(33)/2)/(-2/(-3/2 + sqrt(33)/2) + 2/(-sqrt(33)/2 - 3/2)),                                          2*exp(5/2 - sqrt(33)/2)/((-sqrt(33)/2 - 3/2)*(-2/(-3/2 + sqrt(33)/2) + 2/(-sqrt(33)/2 - 3/2))) - 2*exp(5/2 + sqrt(33)/2)/((-3/2 + sqrt(33)/2)*(-2/(-3/2 + sqrt(33)/2) + 2/(-sqrt(33)/2 - 3/2)))]])
>>> m.eigenvals()
{5/2 - sqrt(33)/2: 1, 5/2 + sqrt(33)/2: 1}

>>> m = Matrix([[1, 2], [2, 4]])
>>> m.exp()
Matrix([
[   4/5 + exp(5)/5, -2/5 + 2*exp(5)/5],
[-2/5 + 2*exp(5)/5,  1/5 + 4*exp(5)/5]])
>>> m.eigenvals()
{5: 1, 0: 1}
```

**Comment by oscarbenjamin** [comment]
This looks good to me except maybe it should be documented somewhere. I think that the matrix exponential might also be undocumented.

**Comment by sylee957** [comment]
One thing I am concerned about is the consistency between the key `simplify` key among many other matrix functions.
Personally, I have an idea of making an option to pass the keyword like `simplify=[simplify, cancel, expand, ...]` and making each simplifying method executed in each call stack.

In this way, it may be possible to throw simplify to intermediate steps, up to user's choice, even more deeply in jordan form computation or nullspace.

However, it may be enigmatic for people who don't read the sympy codebase, and things can be prone to change if internal implementation is changed, but at least it can be useful for people who just want to solve a problem successfully, with trials and errors.

File: sympy/matrices/matrices.py [hunk 2]
```diff
@@ -3200,6 +3213,127 @@ def exp(self):
...
+        l = self[0, 0]
+
+        if l.is_zero:
+            raise MatrixError(
+                'Could not take logarithm or reciprocal for the given '
+                'eigenvalue {}'.format(l))
+
+        bands = {0: log(l)}
+        for i in range(1, size):
+            bands[i] = -((-l) ** -i) / i
```
Comment (jksuom) [inline sympy/matrices/matrices.py]: Why is the exponent negative?

Comment (sylee957) [reply sympy/matrices/matrices.py]: I've only followed the formula (3.8) of https://www.ams.org/journals/proc/1966-017-05/S0002-9939-1966-0202740-6/S0002-9939-1966-0202740-6.pdf

Comment (jksuom) [reply sympy/matrices/matrices.py]: Ahh, the base looked like `-1` on my screen but actually that seems to be `-l`. Perhaps you could define the negative inverse base `b = -1/l` before the loop and then write `b** i`.

**Comment by oscarbenjamin** [comment]
What's the status of this? It looks good to me.

Your last comment discusses simplification and I hope that we get the start of a solution to that in #17247. One thing I think would be good is to be more systematic about how users can control things like this. If we had a consistent approach with functions like `matrix_log`, `matrix_exp`, `matrix_power` etc then I think it would be easier for users to understand what the possibilities are.

**Comment by jksuom** [comment]
Looks good to me too.

**Pull request merged**